packages/std/src/memory.rs

```diff
@@ -116,55 +54,167 @@ unsafe impl RegionSource for Vec<u8> {
+}
+
+pub trait Ownership: sealed::Sealed + 'static {}
+
+impl<T> Ownership for T where T: sealed::Sealed {}
+
+pub struct Owned {}
+
+pub struct Borrowed {}
+
+/// Describes some data allocated in Wasm's linear memory.
+/// A pointer to an instance of this can be returned over FFI boundaries.
+///
+/// This struct is crate internal since the cosmwasm-vm defines the same type independently.
+#[repr(C)]
+pub struct Region<T: Ownership> {
+    /// The beginning of the region expressed as bytes from the beginning of the linear memory
+    pub offset: u32,
+    /// The number of bytes available in this region
+    pub capacity: u32,
+    /// The number of bytes used in this region
+    pub length: u32,
+
+    _marker: PhantomData<T>,
+}
+
+impl Region<Owned> {
+    /// Reconstruct a region from a raw pointer pointing to a `Box<Region>`.
+    /// You'll want to use this when you received a region from the VM and want to dereference its contents.
+    ///
+    /// # Safety
+    ///
+    /// - The pointer must not be null
+    /// - The pointer must be heap allocated
+    /// - This region must point to a valid memory region
+    /// - The memory region this region points to must be heap allocated as well
+    pub unsafe fn from_heap_ptr(ptr: *mut Self) -> Box<Self> {
+        assert!(!ptr.is_null(), "Region pointer is null");
+        Box::from_raw(ptr)
+    }
+
+    /// Construct a new empty region with *at least* a capacity of what you passed in and a length of 0
+    pub fn with_capacity(cap: usize) -> Self {
+        let data = Vec::with_capacity(cap);
+        let region = Self::from_data(data);
+        region
+    }
+
+    pub fn into_inner(self) -> Vec<u8> {
+        let vector = unsafe {
+            Vec::from_raw_parts(
+                self.offset as *mut u8,
+                self.length as usize,
+                self.capacity as usize,
+            )
+        };
+        mem::forget(self);
+        vector
+    }
+}
+
+impl<O> Region<O>
 where
-    T: RegionSource,
+    O: Ownership,
 {
-    fn ptr(&self) -> *const u8 {
-        (**self).ptr()
+    /// Construct a new region from any kind of data that can be turned into a region
+    pub fn from_data<S>(data: S) -> Self
+    where
+        S: RegionSource<Ownership = O>,
+    {
+        // Well, this technically violates pointer provenance rules.
+        // But there isn't a stable API for it, so that's the best we can do, I guess.
+        let region = Region {
+            offset: u32::try_from(data.ptr() as usize).expect("pointer doesn't fit in u32"),
+            capacity: u32::try_from(data.capacity()).expect("capacity doesn't fit in u32"),
+            length: u32::try_from(data.len()).expect("length doesn't fit in u32"),
+
+            _marker: PhantomData,
+        };
+
+        // We gonna forget this.. as a safety measure..
+        // If we didn't do this and the `RegionSource` was a `Vec` we would deallocate it and that's BAD
+        mem::forget(data);
```

Comment:
Wouldn't it be cleaner to have dedicated constructors for owned and borroed sources? Reviewling this, I have no idea if data is `Vec<u8>`, `&Vec<u8>` or `&[u8]`.


Maybe we don't even need the `RegionSource` abstraction anymore